{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keysight Technologies,E4980AL,MY54305061,B.07.01\n",
      "\n",
      "RX\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pyvisa\n",
    "\n",
    "rm = pyvisa.ResourceManager()\n",
    "rm.list_resources()\n",
    "inst = rm.open_resource('GPIB0::17::INSTR')\n",
    "\n",
    "inst.write(\"*IDN?\")\n",
    "print(inst.read())\n",
    "\n",
    "inst.write(\":CORR:LOAD:TYPE RX\")\n",
    "inst.write(\":CORR:LOAD:TYPE?\")\n",
    "print(inst.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymeasure.instruments.agilent import AgilentE4980\n",
    "from pymeasure.experiment.experiment import Experiment\n",
    "from E4980AL import E4980ALProcedure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connect to the instrument and get handle\n",
    "e4980a = AgilentE4980(\"GPIB0::17::INSTR\")\n",
    "e4980a.ac_voltage = 0.022\n",
    "e4980a.mode = \"RX\"\n",
    "e4980a.aperture('short', averages=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12:19:15 AM: Using data file: C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\tmpksdsyxx1 (root, INFO)\n",
      "12:19:15 AM: Using data file: C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\tmpksdsyxx1 (root, INFO)\n",
      "12:19:15 AM: Using data file: C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\tmpksdsyxx1 (root, INFO)\n",
      "12:19:15 AM: Using data file: C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\tmpksdsyxx1 (root, INFO)\n",
      "12:19:15 AM: Set up Results (root, INFO)\n",
      "12:19:15 AM: Set up Results (root, INFO)\n",
      "12:19:15 AM: Set up Results (root, INFO)\n",
      "12:19:15 AM: Set up Results (root, INFO)\n",
      "12:19:15 AM: Create worker (root, INFO)\n",
      "12:19:15 AM: Create worker (root, INFO)\n",
      "12:19:15 AM: Create worker (root, INFO)\n",
      "12:19:15 AM: Create worker (root, INFO)\n",
      "12:19:15 AM: Starting worker... (root, INFO)\n",
      "12:19:15 AM: Starting worker... (root, INFO)\n",
      "12:19:15 AM: Starting worker... (root, INFO)\n",
      "12:19:15 AM: Starting worker... (root, INFO)\n",
      "12:19:15 AM: Worker thread started (root, INFO)\n",
      "12:19:15 AM: Worker thread started (root, INFO)\n",
      "12:19:15 AM: Worker thread started (root, INFO)\n",
      "12:19:15 AM: Worker thread started (root, INFO)\n",
      "12:19:15 AM: Worker started running an instance of 'E4980ALProcedure' (root, INFO)\n",
      "12:19:15 AM: Worker started running an instance of 'E4980ALProcedure' (root, INFO)\n",
      "12:19:15 AM: Worker started running an instance of 'E4980ALProcedure' (root, INFO)\n",
      "12:19:15 AM: Worker started running an instance of 'E4980ALProcedure' (root, INFO)\n",
      "12:19:15 AM: Connecting to the instrument (E4980AL, INFO)\n",
      "12:19:15 AM: Connecting to the instrument (E4980AL, INFO)\n",
      "12:19:15 AM: Connecting to the instrument (E4980AL, INFO)\n",
      "12:19:15 AM: Connecting to the instrument (E4980AL, INFO)\n",
      "12:19:16 AM: Initializing Agilent E4980A/AL LCR meter. (pymeasure.instruments.instrument, INFO)\n",
      "12:19:16 AM: Initializing Agilent E4980A/AL LCR meter. (pymeasure.instruments.instrument, INFO)\n",
      "12:19:16 AM: Initializing Agilent E4980A/AL LCR meter. (pymeasure.instruments.instrument, INFO)\n",
      "12:19:16 AM: Initializing Agilent E4980A/AL LCR meter. (pymeasure.instruments.instrument, INFO)\n",
      "12:19:19 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 173, in run\n",
      "    self.procedure.startup()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 34, in startup\n",
      "    self.instrument = AgilentE4980(self.instrument)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 107, in __init__\n",
      "    self.write(\"FORM ASC\")\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\instrument.py\", line 255, in write\n",
      "    self.adapter.write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\adapter.py\", line 86, in write\n",
      "    self._write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\visa.py\", line 134, in _write\n",
      "    self.connection.write(command, **kwargs)\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 197, in write\n",
      "    count = self.write_raw(message.encode(enco))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 157, in write_raw\n",
      "    return self.visalib.write(self.session, message)[0]\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\functions.py\", line 2796, in write\n",
      "    ret = library.viWrite(session, data, len(data), byref(return_count))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\highlevel.py\", line 226, in _return_handler\n",
      "    return self.handle_return_value(session, ret_value)  # type: ignore\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\highlevel.py\", line 251, in handle_return_value\n",
      "    raise errors.VisaIOError(rv)\n",
      "pyvisa.errors.VisaIOError: VI_ERROR_TMO (-1073807339): Timeout expired before operation completed.\n",
      "12:19:19 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 173, in run\n",
      "    self.procedure.startup()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 34, in startup\n",
      "    self.instrument = AgilentE4980(self.instrument)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 107, in __init__\n",
      "    self.write(\"FORM ASC\")\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\instrument.py\", line 255, in write\n",
      "    self.adapter.write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\adapter.py\", line 86, in write\n",
      "    self._write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\visa.py\", line 134, in _write\n",
      "    self.connection.write(command, **kwargs)\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 197, in write\n",
      "    count = self.write_raw(message.encode(enco))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 157, in write_raw\n",
      "    return self.visalib.write(self.session, message)[0]\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\functions.py\", line 2796, in write\n",
      "    ret = library.viWrite(session, data, len(data), byref(return_count))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\highlevel.py\", line 226, in _return_handler\n",
      "    return self.handle_return_value(session, ret_value)  # type: ignore\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\highlevel.py\", line 251, in handle_return_value\n",
      "    raise errors.VisaIOError(rv)\n",
      "pyvisa.errors.VisaIOError: VI_ERROR_TMO (-1073807339): Timeout expired before operation completed.\n",
      "12:19:19 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 173, in run\n",
      "    self.procedure.startup()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 34, in startup\n",
      "    self.instrument = AgilentE4980(self.instrument)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 107, in __init__\n",
      "    self.write(\"FORM ASC\")\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\instrument.py\", line 255, in write\n",
      "    self.adapter.write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\adapter.py\", line 86, in write\n",
      "    self._write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\visa.py\", line 134, in _write\n",
      "    self.connection.write(command, **kwargs)\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 197, in write\n",
      "    count = self.write_raw(message.encode(enco))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 157, in write_raw\n",
      "    return self.visalib.write(self.session, message)[0]\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\functions.py\", line 2796, in write\n",
      "    ret = library.viWrite(session, data, len(data), byref(return_count))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\highlevel.py\", line 226, in _return_handler\n",
      "    return self.handle_return_value(session, ret_value)  # type: ignore\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\highlevel.py\", line 251, in handle_return_value\n",
      "    raise errors.VisaIOError(rv)\n",
      "pyvisa.errors.VisaIOError: VI_ERROR_TMO (-1073807339): Timeout expired before operation completed.\n",
      "12:19:19 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 173, in run\n",
      "    self.procedure.startup()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 34, in startup\n",
      "    self.instrument = AgilentE4980(self.instrument)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 107, in __init__\n",
      "    self.write(\"FORM ASC\")\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\instrument.py\", line 255, in write\n",
      "    self.adapter.write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\adapter.py\", line 86, in write\n",
      "    self._write(command, **kwargs)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\adapters\\visa.py\", line 134, in _write\n",
      "    self.connection.write(command, **kwargs)\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 197, in write\n",
      "    count = self.write_raw(message.encode(enco))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\resources\\messagebased.py\", line 157, in write_raw\n",
      "    return self.visalib.write(self.session, message)[0]\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\functions.py\", line 2796, in write\n",
      "    ret = library.viWrite(session, data, len(data), byref(return_count))\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\ctwrapper\\highlevel.py\", line 226, in _return_handler\n",
      "    return self.handle_return_value(session, ret_value)  # type: ignore\n",
      "  File \"C:\\Users\\Administrator\\AppData\\Roaming\\Python\\Python39\\site-packages\\pyvisa\\highlevel.py\", line 251, in handle_return_value\n",
      "    raise errors.VisaIOError(rv)\n",
      "pyvisa.errors.VisaIOError: VI_ERROR_TMO (-1073807339): Timeout expired before operation completed.\n",
      "12:19:25 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:25 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:25 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:25 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:35 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:35 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:35 AM: Timeout, no data received for liveplot (root, WARNING)\n",
      "12:19:35 AM: Timeout, no data received for liveplot (root, WARNING)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute '_xorig'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[8], line 33\u001b[0m\n\u001b[0;32m     31\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m     32\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 33\u001b[0m         experiment\u001b[39m.\u001b[39;49mupdate_line(ax1, hl1, \u001b[39m'\u001b[39;49m\u001b[39mFrequency\u001b[39;49m\u001b[39m'\u001b[39;49m,\u001b[39m'\u001b[39;49m\u001b[39mZ\u001b[39;49m\u001b[39m\\'\u001b[39;49;00m\u001b[39m'\u001b[39;49m)\n\u001b[0;32m     34\u001b[0m         pl\u001b[39m.\u001b[39mpause(\u001b[39m0.1\u001b[39m)\n\u001b[0;32m     35\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\experiment.py:214\u001b[0m, in \u001b[0;36mExperiment.update_line\u001b[1;34m(self, ax, hl, xname, yname)\u001b[0m\n\u001b[0;32m    212\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mupdate_line\u001b[39m(\u001b[39mself\u001b[39m, ax, hl, xname, yname):\n\u001b[0;32m    213\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Update a line in a matplotlib graph with new data.\"\"\"\u001b[39;00m\n\u001b[1;32m--> 214\u001b[0m     \u001b[39mdel\u001b[39;00m hl\u001b[39m.\u001b[39m_xorig, hl\u001b[39m.\u001b[39m_yorig\n\u001b[0;32m    215\u001b[0m     hl\u001b[39m.\u001b[39mset_xdata(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_data[xname])\n\u001b[0;32m    216\u001b[0m     hl\u001b[39m.\u001b[39mset_ydata(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_data[yname])\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'list' object has no attribute '_xorig'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAi4AAAGiCAYAAADA0E3hAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAdM0lEQVR4nO3df2zXhZ348VdpaavetYswKwoy2OlkI3OjBEZJz5ynXdC48MdFFi+Cnkuu2XaInN5gXGQYk2ZbZjK3wX6BZgl6REXPPzhn/9gUxWwnV5ZlkLgIt8JWJMXYos4y4P39wy+91Bbl8+HTHy99PJLPH337ftMXG+9Xnp/PBz6tKoqiCACABCaN9wAAAGdLuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGmUHC7PPfdc3HjjjXHJJZdEVVVVPPnkk+97zbPPPhvNzc1RX18fs2fPjh/+8IflzAokZW8AlVJyuLz55ptx1VVXxfe///2zOv/AgQNx/fXXR2tra3R1dcXXv/71WLlyZTz++OMlDwvkZG8AlVJ1Lj9ksaqqKp544olYunTpGc/52te+Fk899VTs27dv8Fh7e3v85je/iRdffLHcbw0kZW8A56JmtL/Biy++GG1tbUOOff7zn4/NmzfHX/7yl5g8efKwawYGBmJgYGDw61OnTsVrr70WU6ZMiaqqqtEeGXiXoiji2LFjcckll8SkSaP/V+PsDfhgGI3dMerhcvjw4WhqahpyrKmpKU6cOBG9vb0xbdq0Ydd0dHTEhg0bRns0oEQHDx6M6dOnj/r3sTfgg6WSu2PUwyUihj3bOf3u1JmeBa1duzZWr149+HVfX19cdtllcfDgwWhoaBi9QYER9ff3x4wZM+Kv//qvx+x72huQ32jsjlEPl4svvjgOHz485NiRI0eipqYmpkyZMuI1dXV1UVdXN+x4Q0ODBQTjaKzecrE34IOlkrtj1N+sXrRoUXR2dg459swzz8T8+fNHfJ8awN4AzqTkcHnjjTdiz549sWfPnoh4558t7tmzJ7q7uyPinZdrly9fPnh+e3t7/OEPf4jVq1fHvn37YsuWLbF58+a46667KvM7ACY8ewOomKJEv/jFL4qIGPZYsWJFURRFsWLFiuLqq68ecs0vf/nL4rOf/WxRW1tbfOxjHys2bdpU0vfs6+srIqLo6+srdVygAs71HrQ34MNpNO7Dc/ocl7HS398fjY2N0dfX571qGAcZ78GMM8MHzWjch35WEQCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKRRVrhs3LgxZs2aFfX19dHc3Bw7d+58z/O3bt0aV111VZx//vkxbdq0uO222+Lo0aNlDQzkZG8AlVByuGzbti1WrVoV69ati66urmhtbY0lS5ZEd3f3iOc///zzsXz58rj99tvjd7/7XTz66KPx3//93/GlL33pnIcHcrA3gIopSrRgwYKivb19yLErr7yyWLNmzYjnf/vb3y5mz5495NgDDzxQTJ8+/ay/Z19fXxERRV9fX6njAhVwrvegvQEfTqNxH5b0isvx48dj9+7d0dbWNuR4W1tb7Nq1a8RrWlpa4tChQ7Fjx44oiiJeffXVeOyxx+KGG2444/cZGBiI/v7+IQ8gJ3sDqKSSwqW3tzdOnjwZTU1NQ443NTXF4cOHR7ympaUltm7dGsuWLYva2tq4+OKL4yMf+Uh873vfO+P36ejoiMbGxsHHjBkzShkTmEDsDaCSyvrLuVVVVUO+Lopi2LHT9u7dGytXrox77rkndu/eHU8//XQcOHAg2tvbz/jrr127Nvr6+gYfBw8eLGdMYAKxN4BKqCnl5KlTp0Z1dfWwZ0lHjhwZ9mzqtI6Ojli8eHHcfffdERHx6U9/Oi644IJobW2N++67L6ZNmzbsmrq6uqirqytlNGCCsjeASirpFZfa2tpobm6Ozs7OIcc7OzujpaVlxGveeuutmDRp6Leprq6OiHeecQEfbPYGUEklv1W0evXq+OlPfxpbtmyJffv2xZ133hnd3d2DL+GuXbs2li9fPnj+jTfeGNu3b49NmzbF/v3744UXXoiVK1fGggUL4pJLLqnc7wSYsOwNoFJKeqsoImLZsmVx9OjRuPfee6Onpyfmzp0bO3bsiJkzZ0ZERE9Pz5DPZrj11lvj2LFj8f3vfz/+9V//NT7ykY/ENddcE9/85jcr97sAJjR7A6iUqiLB6679/f3R2NgYfX190dDQMN7jwIdOxnsw48zwQTMa96GfVQQApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABplBUuGzdujFmzZkV9fX00NzfHzp073/P8gYGBWLduXcycOTPq6uri4x//eGzZsqWsgYGc7A2gEmpKvWDbtm2xatWq2LhxYyxevDh+9KMfxZIlS2Lv3r1x2WWXjXjNTTfdFK+++mps3rw5/uZv/iaOHDkSJ06cOOfhgRzsDaBSqoqiKEq5YOHChTFv3rzYtGnT4LE5c+bE0qVLo6OjY9j5Tz/9dHzxi1+M/fv3x4UXXljWkP39/dHY2Bh9fX3R0NBQ1q8BlO9c70F7Az6cRuM+LOmtouPHj8fu3bujra1tyPG2trbYtWvXiNc89dRTMX/+/PjWt74Vl156aVxxxRVx1113xZ///Oczfp+BgYHo7+8f8gBysjeASirpraLe3t44efJkNDU1DTne1NQUhw8fHvGa/fv3x/PPPx/19fXxxBNPRG9vb3z5y1+O11577YzvV3d0dMSGDRtKGQ2YoOwNoJLK+su5VVVVQ74uimLYsdNOnToVVVVVsXXr1liwYEFcf/31cf/998dDDz10xmdPa9eujb6+vsHHwYMHyxkTmEDsDaASSnrFZerUqVFdXT3sWdKRI0eGPZs6bdq0aXHppZdGY2Pj4LE5c+ZEURRx6NChuPzyy4ddU1dXF3V1daWMBkxQ9gZQSSW94lJbWxvNzc3R2dk55HhnZ2e0tLSMeM3ixYvjT3/6U7zxxhuDx15++eWYNGlSTJ8+vYyRgUzsDaCSSn6raPXq1fHTn/40tmzZEvv27Ys777wzuru7o729PSLeebl2+fLlg+fffPPNMWXKlLjtttti79698dxzz8Xdd98d//RP/xTnnXde5X4nwIRlbwCVUvLnuCxbtiyOHj0a9957b/T09MTcuXNjx44dMXPmzIiI6Onpie7u7sHz/+qv/io6OzvjX/7lX2L+/PkxZcqUuOmmm+K+++6r3O8CmNDsDaBSSv4cl/Hg8xhgfGW8BzPODB804/45LgAA40m4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSKCtcNm7cGLNmzYr6+vpobm6OnTt3ntV1L7zwQtTU1MRnPvOZcr4tkJi9AVRCyeGybdu2WLVqVaxbty66urqitbU1lixZEt3d3e95XV9fXyxfvjz+/u//vuxhgZzsDaBSqoqiKEq5YOHChTFv3rzYtGnT4LE5c+bE0qVLo6Oj44zXffGLX4zLL788qqur48knn4w9e/ac8dyBgYEYGBgY/Lq/vz9mzJgRfX190dDQUMq4QAX09/dHY2Nj2fegvQEfTue6O0ZS0isux48fj927d0dbW9uQ421tbbFr164zXvfggw/GK6+8EuvXrz+r79PR0RGNjY2DjxkzZpQyJjCB2BtAJZUULr29vXHy5MloamoacrypqSkOHz484jW///3vY82aNbF169aoqak5q++zdu3a6OvrG3wcPHiwlDGBCcTeACrp7DbCu1RVVQ35uiiKYcciIk6ePBk333xzbNiwIa644oqz/vXr6uqirq6unNGACcreACqhpHCZOnVqVFdXD3uWdOTIkWHPpiIijh07Fi+99FJ0dXXFV7/61YiIOHXqVBRFETU1NfHMM8/ENddccw7jAxOdvQFUUklvFdXW1kZzc3N0dnYOOd7Z2RktLS3Dzm9oaIjf/va3sWfPnsFHe3t7fOITn4g9e/bEwoULz216YMKzN4BKKvmtotWrV8ctt9wS8+fPj0WLFsWPf/zj6O7ujvb29oh4533mP/7xj/Gzn/0sJk2aFHPnzh1y/UUXXRT19fXDjgMfXPYGUCklh8uyZcvi6NGjce+990ZPT0/MnTs3duzYETNnzoyIiJ6envf9bAbgw8XeACql5M9xGQ+j8e/AgbOX8R7MODN80Iz757gAAIwn4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASKOscNm4cWPMmjUr6uvro7m5OXbu3HnGc7dv3x7XXXddfPSjH42GhoZYtGhR/PznPy97YCAnewOohJLDZdu2bbFq1apYt25ddHV1RWtrayxZsiS6u7tHPP+5556L6667Lnbs2BG7d++Ov/u7v4sbb7wxurq6znl4IAd7A6iUqqIoilIuWLhwYcybNy82bdo0eGzOnDmxdOnS6OjoOKtf41Of+lQsW7Ys7rnnnhH/+8DAQAwMDAx+3d/fHzNmzIi+vr5oaGgoZVygAvr7+6OxsbHse9DegA+nc90dIynpFZfjx4/H7t27o62tbcjxtra22LVr11n9GqdOnYpjx47FhRdeeMZzOjo6orGxcfAxY8aMUsYEJhB7A6ikksKlt7c3Tp48GU1NTUOONzU1xeHDh8/q1/jOd74Tb775Ztx0001nPGft2rXR19c3+Dh48GApYwITiL0BVFJNORdVVVUN+booimHHRvLII4/EN77xjfjP//zPuOiii854Xl1dXdTV1ZUzGjBB2RtAJZQULlOnTo3q6uphz5KOHDky7NnUu23bti1uv/32ePTRR+Paa68tfVIgJXsDqKSS3iqqra2N5ubm6OzsHHK8s7MzWlpaznjdI488Erfeems8/PDDccMNN5Q3KZCSvQFUUslvFa1evTpuueWWmD9/fixatCh+/OMfR3d3d7S3t0fEO+8z//GPf4yf/exnEfHO8lm+fHl897vfjc997nODz7rOO++8aGxsrOBvBZio7A2gUkoOl2XLlsXRo0fj3nvvjZ6enpg7d27s2LEjZs6cGRERPT09Qz6b4Uc/+lGcOHEivvKVr8RXvvKVweMrVqyIhx566Nx/B8CEZ28AlVLy57iMh9H4d+DA2ct4D2acGT5oxv1zXAAAxpNwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGsIFAEhDuAAAaQgXACAN4QIApCFcAIA0hAsAkIZwAQDSEC4AQBrCBQBIQ7gAAGkIFwAgDeECAKQhXACANIQLAJCGcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkUVa4bNy4MWbNmhX19fXR3NwcO3fufM/zn3322Whubo76+vqYPXt2/PCHPyxrWCAvewOohJLDZdu2bbFq1apYt25ddHV1RWtrayxZsiS6u7tHPP/AgQNx/fXXR2tra3R1dcXXv/71WLlyZTz++OPnPDyQg70BVEpVURRFKRcsXLgw5s2bF5s2bRo8NmfOnFi6dGl0dHQMO/9rX/taPPXUU7Fv377BY+3t7fGb3/wmXnzxxRG/x8DAQAwMDAx+3dfXF5dddlkcPHgwGhoaShkXqID+/v6YMWNGvP7669HY2Fjy9fYGfDid6+4YUVGCgYGBorq6uti+ffuQ4ytXriz+9m//dsRrWltbi5UrVw45tn379qKmpqY4fvz4iNesX7++iAgPD48J9njllVdKWRn2hoeHRxFR3u44k5ooQW9vb5w8eTKampqGHG9qaorDhw+PeM3hw4dHPP/EiRPR29sb06ZNG3bN2rVrY/Xq1YNfv/766zFz5szo7u6uXLGNstOVmenZnpnHRsaZT796ceGFF5Z8rb1x9jL+2YjIObeZx8a57I4zKSlcTquqqhrydVEUw4693/kjHT+trq4u6urqhh1vbGxM83/WaQ0NDWYeA2YeG5Mmlf8PEe2Ns5fxz0ZEzrnNPDbOZXcM+7VKOXnq1KlRXV097FnSkSNHhj07Ou3iiy8e8fyampqYMmVKieMC2dgbQCWVFC61tbXR3NwcnZ2dQ453dnZGS0vLiNcsWrRo2PnPPPNMzJ8/PyZPnlziuEA29gZQUaX+pZj/+I//KCZPnlxs3ry52Lt3b7Fq1ariggsuKP73f/+3KIqiWLNmTXHLLbcMnr9///7i/PPPL+68885i7969xebNm4vJkycXjz322Fl/z7fffrtYv3598fbbb5c67rgx89gw89g415ntjbOTceaiyDm3mcfGaMxccrgURVH84Ac/KGbOnFnU1tYW8+bNK5599tnB/7ZixYri6quvHnL+L3/5y+Kzn/1sUVtbW3zsYx8rNm3adE5DA/nYG0AllPw5LgAA48XPKgIA0hAuAEAawgUASEO4AABpTJhwyfgj70uZefv27XHdddfFRz/60WhoaIhFixbFz3/+8zGc9h2l/u982gsvvBA1NTXxmc98ZnQHHEGpMw8MDMS6deti5syZUVdXFx//+Mdjy5YtYzTtO0qdeevWrXHVVVfF+eefH9OmTYvbbrstjh49OkbTRjz33HNx4403xiWXXBJVVVXx5JNPvu812e7BiHwz2xvly7g3InLtjnHbG+P9z5qK4v8+4+EnP/lJsXfv3uKOO+4oLrjgguIPf/jDiOef/oyHO+64o9i7d2/xk5/8pOTPeBjrme+4447im9/8ZvHrX/+6ePnll4u1a9cWkydPLv7nf/5nws582uuvv17Mnj27aGtrK6666qqxGfb/K2fmL3zhC8XChQuLzs7O4sCBA8WvfvWr4oUXXpiwM+/cubOYNGlS8d3vfrfYv39/sXPnzuJTn/pUsXTp0jGbeceOHcW6deuKxx9/vIiI4oknnnjP8zPegxlntjfKk3FvFEW+3TFee2NChMuCBQuK9vb2IceuvPLKYs2aNSOe/2//9m/FlVdeOeTYP//zPxef+9znRm3Gdyt15pF88pOfLDZs2FDp0c6o3JmXLVtW/Pu//3uxfv36MV9Apc78X//1X0VjY2Nx9OjRsRhvRKXO/O1vf7uYPXv2kGMPPPBAMX369FGb8b2czQLKeA9mnHkk9sb7y7g3iiL37hjLvTHubxUdP348du/eHW1tbUOOt7W1xa5du0a85sUXXxx2/uc///l46aWX4i9/+cuozXpaOTO/26lTp+LYsWMV/YmZ76XcmR988MF45ZVXYv369aM94jDlzPzUU0/F/Pnz41vf+lZceumlccUVV8Rdd90Vf/7zn8di5LJmbmlpiUOHDsWOHTuiKIp49dVX47HHHosbbrhhLEYuS8Z7MOPM72ZvvL+MeyPiw7E7KnUPlvXToStprH7kfSWVM/O7fec734k333wzbrrpptEYcZhyZv79738fa9asiZ07d0ZNzdj/USln5v3798fzzz8f9fX18cQTT0Rvb298+ctfjtdee21M3q8uZ+aWlpbYunVrLFu2LN5+++04ceJEfOELX4jvfe97oz5vuTLegxlnfjd74/1l3BsRH47dUal7cNxfcTlttH/k/WgodebTHnnkkfjGN74R27Zti4suumi0xhvR2c588uTJuPnmm2PDhg1xxRVXjNV4Iyrlf+dTp05FVVVVbN26NRYsWBDXX3993H///fHQQw+N6bOnUmbeu3dvrFy5Mu65557YvXt3PP3003HgwIFob28fi1HLlvEezDjzafZGaTLujYgP/u6oxD047q+4ZPyR9+XMfNq2bdvi9ttvj0cffTSuvfba0RxziFJnPnbsWLz00kvR1dUVX/3qVyPinZu7KIqoqamJZ555Jq655poJNXNExLRp0+LSSy+NxsbGwWNz5syJoiji0KFDcfnll0+4mTs6OmLx4sVx9913R0TEpz/96bjggguitbU17rvvvlF/JaAcGe/BjDOfZm+M3swR4783Ij4cu6NS9+C4v+KS8UfelzNzxDvPmG699dZ4+OGHx/w9yFJnbmhoiN/+9rexZ8+ewUd7e3t84hOfiD179sTChQsn3MwREYsXL44//elP8cYbbwwee/nll2PSpEkxffr0UZ03oryZ33rrrZg0aeitWF1dHRH/92xkosl4D2acOcLeGO2ZI8Z/b0R8OHZHxe7Bkv4q7ygZjx95P9YzP/zww0VNTU3xgx/8oOjp6Rl8vP766xN25ncbj38dUOrMx44dK6ZPn178wz/8Q/G73/2uePbZZ4vLL7+8+NKXvjRhZ37wwQeLmpqaYuPGjcUrr7xSPP/888X8+fOLBQsWjNnMx44dK7q6uoqurq4iIor777+/6OrqGvxnmB+EezDjzPZGeTLujXLmHu/dMV57Y0KES1Hk/JH3pcx89dVXFxEx7LFixYoJO/O7jccCKorSZ963b19x7bXXFuedd14xffr0YvXq1cVbb701oWd+4IEHik9+8pPFeeedV0ybNq34x3/8x+LQoUNjNu8vfvGL9/zz+UG4B4si38z2Rvky7o2iyLU7xmtvVBXFBHw9CQBgBOP+d1wAAM6WcAEA0hAuAEAawgUASEO4AABpCBcAIA3hAgCkIVwAgDSECwCQhnABANIQLgBAGv8P6EDt3Xw6P4gAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12:20:10 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 176, in run\n",
      "    self.procedure.execute()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 47, in execute\n",
      "    results=self.instrument.freq_sweep(frequencies)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in freq_sweep\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in <listcomp>\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "IndexError: list index out of range\n",
      "12:20:10 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 176, in run\n",
      "    self.procedure.execute()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 47, in execute\n",
      "    results=self.instrument.freq_sweep(frequencies)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in freq_sweep\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in <listcomp>\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "IndexError: list index out of range\n",
      "12:20:10 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 176, in run\n",
      "    self.procedure.execute()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 47, in execute\n",
      "    results=self.instrument.freq_sweep(frequencies)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in freq_sweep\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in <listcomp>\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "IndexError: list index out of range\n",
      "12:20:10 AM: Worker caught an error on <E4980ALProcedure(status=Running,parameters_are_set=True)> (root, ERROR)\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\experiment\\workers.py\", line 176, in run\n",
      "    self.procedure.execute()\n",
      "  File \"c:\\Users\\Administrator\\Documents\\GitHub\\Keysight_E4980AL_impedance\\E4980AL.py\", line 47, in execute\n",
      "    results=self.instrument.freq_sweep(frequencies)\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in freq_sweep\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "  File \"c:\\ProgramData\\Anaconda3\\lib\\site-packages\\pymeasure\\instruments\\agilent\\agilentE4980.py\", line 141, in <listcomp>\n",
      "    a_data = [measured[_] for _ in range(0, 4 * len(freq_list), 4)]\n",
      "IndexError: list index out of range\n"
     ]
    }
   ],
   "source": [
    "# conduct \"fast\" sweep\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "datafile='test.csv'\n",
    "\n",
    "sweep_type='LIN'\n",
    "start_freq=20\n",
    "stop_freq=100000\n",
    "sweep_reverse=True\n",
    "num_points=15\n",
    "if sweep_type=='LIN':\n",
    "    frequencies = np.linspace(start_freq, stop_freq, num_points)\n",
    "elif sweep_type=='LOG':\n",
    "    frequencies = np.logspace(np.log10(start_freq), np.log10(stop_freq), num_points)\n",
    "if sweep_reverse:\n",
    "    frequencies = frequencies[::-1]\n",
    "\n",
    "title=\"E4980AL Test\"\n",
    "procedure = E4980ALProcedure()\n",
    "experiment = Experiment(title, procedure)\n",
    "experiment.start()\n",
    "\n",
    "import pylab as pl\n",
    "ax1 = pl.subplot(121)\n",
    "experiment.plot('Frequency','Z\\'',ax=ax1)\n",
    "hl1,_ = ax1.get_legend_handles_labels()\n",
    "ax2 = pl.subplot(122)\n",
    "experiment.plot('Frequency','Z\\\"',ax=ax2)\n",
    "hl2,_ = ax2.get_legend_handles_labels()\n",
    "while True:\n",
    "    try:\n",
    "        experiment.update_line(ax1, hl1[0], 'Frequency','Z\\'')\n",
    "        pl.pause(0.1)\n",
    "    except KeyboardInterrupt:\n",
    "        break\n",
    "    \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# conduct \"slow\" sweep\n",
    "import numpy as np\n",
    "sweep_type='LIN'\n",
    "start_freq=20\n",
    "stop_freq=100000\n",
    "sweep_reverse=True\n",
    "num_points=15\n",
    "if sweep_type=='LIN':\n",
    "    frequencies = np.linspace(start_freq, stop_freq, num_points)\n",
    "elif sweep_type=='LOG':\n",
    "    frequencies = np.logspace(np.log10(start_freq), np.log10(stop_freq), num_points)\n",
    "if sweep_reverse:\n",
    "    frequencies = frequencies[::-1]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Keysight Technologies,E4980AL,MY54305061,B.07.01\n",
      "\n"
     ]
    }
   ],
   "source": [
    "read_value=e4980a.ask(\"*IDN?\")\n",
    "print(read_value)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
